container-search/src/main/java/com/yahoo/search/dispatch/searchcluster/SearchGroupsImpl.java

```diff
@@ -23,23 +26,38 @@ public SearchGroupsImpl(Map<Integer, Group> groups, double minActivedocsPercenta
     @Override public int size() { return groups.size(); }
 
     @Override
-    public boolean isPartialGroupCoverageSufficient(Collection<Node> nodes) {
-        if (size() == 1)
-            return true;
-        long activeDocuments = nodes.stream().mapToLong(Node::getActiveDocuments).sum();
-        return isGroupCoverageSufficient(activeDocuments, medianDocumentsPerGroup());
+    public boolean isPartialGroupCoverageSufficient(boolean currentIsGroupCoverageSufficient, Collection<Node> nodes) {
+        if (size() == 1) return true;
+        long groupDocumentCount = nodes.stream().mapToLong(Node::getActiveDocuments).sum();
+        return isGroupCoverageSufficient(currentIsGroupCoverageSufficient,
+                                         groupDocumentCount, medianDocumentCount(), maxDocumentCount());
     }
 
-    public boolean isGroupCoverageSufficient(long activeDocuments, long medianDocuments) {
-        if (medianDocuments <= 0) return true;
-        double documentCoverage = 100.0 * (double) activeDocuments / medianDocuments;
-        return documentCoverage >= minActivedocsPercentage;
+    public boolean isGroupCoverageSufficient(boolean currentIsGroupCoverageSufficient,
+                                             long groupDocumentCount, long medianDocumentCount, long maxDocumentCount) {
+        if (medianDocumentCount <= 0) return true;
+        if (currentIsGroupCoverageSufficient) {
+            // To take a group *out of* rotation, require that it has less active documents than the median.
+            // This avoids scenarios where incorrect accounting in a single group takes all other groups offline.
+            double documentCoverage = 100.0 * (double) groupDocumentCount / medianDocumentCount;
+            return documentCoverage >= minActiveDocsPercentage;
+        }
+        else {
+            // to put a group *in* rotation, require that it has as many documents as the largest group,
```

Comment:
I wonder if there are situations where this will prevent groups that should have been included to join.
 Need to let this one mature a bit.

What about median(of all groups above average) ?

Reply:
That would be better in the case where you have one group with anomalously much data, at least three existing groups, and some new groups added. 

But if you had one group and are adding N new, then the larger N gets, the closer the average gets towards what those new groups have indexed so far on average, and if more than 1 of the new ones happens to go enough above the others to go above average, then 1 or more of them will be added prematurely.

Reply:
I guess either of these solutions are fine, but I suggest we run them by @vekterli first.

Reply:
One concern is that using "sticky" state (i.e. "does the group already have sufficient coverage?") as an input to the group coverage algorithm introduces an asymmetry between the decisions made by a container that has been up for some time versus one that has just started up. This because internal state is transient and groups are considered _down_ initially upon container startup.

Hypothetical scenario: the cluster is in some edge case where all but 1 group has coverage that is < 97% of max active but >= median active. If the container was alive when these groups had >= 97% coverage, they will _remain_ available for serving. If the container then restarts, this information is lost and the groups will _not_ be available for serving until they have reached >= 97% coverage once more.

I don't see this as something that should happen during normal operation, and these changes should make the behavior when adding fresh groups much better than it is today. So we'll probably have to observe it in practice for a bit before making any adjustments (related; should this behavior be guarded by a feature flag?).